src/ElfReader.cc

```diff
@@ -317,7 +380,31 @@ ScopedFd ElfFileReader::open_debug_file(const std::string& elf_file_name) {
   string debug_path = "/usr/lib/debug/";
   debug_path +=
     elf_file_name.substr(0, last_slash) + '/' + debuglink.filename.data();
-  return ScopedFd(debug_path.c_str(), O_RDONLY);
+  ScopedFd debug_fd(debug_path.c_str(), O_RDONLY);
+  if (!debug_fd.is_open()) {
+    return ScopedFd();
+  }
+
+  uint32_t crc = 0xffffffff;
+  while (true) {
+    unsigned char buf[4096];
+    ssize_t ret = ::read(debug_fd.get(), buf, sizeof(buf));
+    if (ret < 0) {
+      if (errno != EINTR) {
+        LOG(debug) << "Error reading " << debug_path;
+        return ScopedFd();
+      }
+    } else if (ret == 0) {
+      break;
+    } else {
+      crc = crc32(crc, buf, ret);
+    }
+  }
+
+  if ((crc ^ 0xffffffff) == debuglink.crc) {
+    return debug_fd;
+  }
+  return ScopedFd();
```

Comment:
"the debug link specifies a 32-bit Cyclic Redundancy Check (CRC) checksum for
the debug file, which GDB uses to validate that the executable and the debug
file came from the same build."
FWIW breakpad also checks the CRC.

I assume this is to protect against the debug file and the text file being out
of sync, on systems where they are in different packages.  It is not necessary
on Gentoo, because the files both come from the same package.

Reading the whole debug file seems expensive, and so there's definitely
a case for skipping this.

[1] https://sourceware.org/gdb/onlinedocs/gdb/Separate-Debug-Files.html